### PR TITLE
If logging as JSON and the type is json.RawMessage log it "as-is"

### PR DESCRIPTION
**Pull request by sfc-gh-jchacon**
If we're already producing JSON and a sub message field is json.RawMessage (i.e. []byte) then emit it as-is.

Otherwise you get [X,Y,Z,..] style which is pretty useless in the log in that form.

### CODE REVIEW

**Comment by pohly** [comment]
I'm not sure about this. How would code know that it is okay to pass a `json.RawMessage` as a value? This is only useful in code which will only ever be used with funcr configured to emit JSON output, which kind of defeats the purpose of go-logr as a general-purpose logging API.

**Comment by sfc-gh-jchacon** [comment]
That's why this is in the funcr specific code that is logging in JSON format. It's scoped directly to that and not in the general case on purpose. 

The reason for it is if you have a json.RawMessage you likely want to see it in string form not [X,Y,Z] form. Converting it to string by casting to string also doesn't work as the pretty printer gets involved at that point and escapes all the JSON.

**Comment by sfc-gh-jchacon** [comment]
The other option would be to add a struct tag which adds an option when logging which says "just stringify this as-is".

i.e.

type Message struct {
  Data json.RawMessage `json:"data` `funcr:"stringify"`
}

So instead of [X,Y,Z] you'd just get the json string in there directly.

Thoughts?

**Comment by pohly** [comment]
> That's why this is in the funcr specific code that is logging in JSON format. It's scoped directly to that and not in the general case on purpose.

Let me rephrase: how will this feature be used in log calls? How does that code know that passing such a special value is okay? Or is this code never meant to be used with any logger other than funcr with JSON output?

**Comment by sfc-gh-jchacon** [comment]
> > That's why this is in the funcr specific code that is logging in JSON format. It's scoped directly to that and not in the general case on purpose.
> 
> Let me rephrase: how will this feature be used in log calls? How does that code know that passing such a special value is okay? Or is this code never meant to be used with any logger other than funcr with JSON output?

Yes, that's the idea. This is specific to funcr and its ability to easily render as JSON instead of KV pairs. 

My working example of this was a gRPC server that was taking a proto.Message and serializing as JSON before passing to OPA for policy evaluation. The OPA system sees the []byte and can treat that as a direct string but we can't log the request in human readable format today without this. i.e. [X,Y,Z] form or I cast the RawMessage to a string but then the pretty printer turns it into escaped JSON which isn't useful either.

File: funcr/funcr.go [hunk 1]
```diff
@@ -500,6 +501,21 @@ func (f Formatter) prettyWithFlags(value interface{}, flags uint32, depth int) s
...
+		// If this is outputing as JSON make sure this isn't really a json.RawMessage.
+		// If so just emit "as-is" and don't pretty it as that will just print
+		// it as [X,Y,Z,...] which isn't terribly useful vs the string form you really want.
+		if f.outputFormat == outputJSON {
+			if rm, ok := value.(json.RawMessage); ok {
+				// If it's empty make sure we emit an empty value as the array style would below.
+				if len(rm) > 0 {
+					buf.Write(rm)
+				} else {
+					buf.WriteByte('{')
```
Comment (thockin) [inline funcr/funcr.go:513]: I wonder if there's a schema problem here - if the consumer of the log expects an int or string or list, and you give them an object, will it cause a problem?  I guess it's a super corner-case, but why not put "null" here instead?

**Comment by thockin** [comment]
Any desire to finish this one?

File: funcr/funcr.go (hunk 1 above)
Comment (sfc-gh-jchacon) [reply funcr/funcr.go:513]: Ahh...true. Let me do that

File: funcr/funcr_test.go [hunk 2]
```diff
@@ -738,6 +751,26 @@ func TestRender(t *testing.T) {
...
+	}, {
+		name:       "byte array not json.RawMessage",
+		args:       makeKV("key", []byte{1, 2, 3, 4}),
+		expectKV:   `"key"=[1,2,3,4]`,
+		expectJSON: `{"key":[1,2,3,4]}`,
+	}, {
+		name:       "json rendering with empty json.RawMessage",
+		args:       makeKV("key", &Trawjson{}),
+		expectKV:   `"key"={"message":[]}`,
+		expectJSON: `{"key":{"message":{}}}`,
```
Comment (thockin) [inline funcr/funcr_test.go:768]: should be "null" ?

File: funcr/funcr_test.go [hunk 3]
```diff
@@ -738,6 +751,26 @@ func TestRender(t *testing.T) {
...
+	}, {
+		name:       "byte array not json.RawMessage",
```
Comment (thockin) [inline funcr/funcr_test.go:770]: this case seems to be a dup of the above?

File: funcr/funcr_test.go (hunk 2 above)
Comment (sfc-gh-jchacon) [reply funcr/funcr_test.go]: Sigh. Done

File: funcr/funcr_test.go (hunk 3 above)
Comment (sfc-gh-jchacon) [reply funcr/funcr_test.go]: Yep. Must have cut/paste. Cleaned up
